[jit] improve recursive scripting error message

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21842 [jit] improve error message on recursive class defs
* **#21841 [jit] improve recursive scripting error message**

We try to grab the __code__ attribute, which doesn't work for classes.
Add a better error message in this case

Differential Revision: [D15857569](https://our.internmc.facebook.com/intern/diff/D15857569)

suo committed:
[jit] improve recursive scripting error message

We try to grab the __code__ attribute, which doesn't work for classes.
Add a better error message in this case

suo committed:
Update on "[jit] improve recursive scripting error message"

[jit] improve recursive scripting error message

We try to grab the __code__ attribute, which doesn't work for classes.
Add a better error message in this case

gh-metadata: pytorch pytorch 21841 gh/suo/62/head

zdevito reviewed (approved):
Seems fine, but the error message doesn't say what the `fn` is if not a function and we got there. Also why is it valid to ask for `__name__` of a random object?

suo committed:
Update on "[jit] improve recursive scripting error message"

[jit] improve recursive scripting error message

We try to grab the __code__ attribute, which doesn't work for classes.
Add a better error message in this case

gh-metadata: pytorch pytorch 21841 gh/suo/62/head

suo committed:
Update on "[jit] improve recursive scripting error message"

[jit] improve recursive scripting error message

We try to grab the __code__ attribute, which doesn't work for classes.
Add a better error message in this case

gh-metadata: pytorch pytorch 21841 gh/suo/62/head
